### PR TITLE
Removed save_every_N sequence size limitiation.

### LINKED ISSUE #5106: Generalize Scan memory/speed trade-off
https://github.com/Theano/Theano/issues/4124 issue was about adding that.

Current implementation have this restriction that we want to lift:
- [ ] The value of `save_every_N` divides the number of steps the scan will run without remainder.

I'm not sure we should lift this limit. What do you think?
- Only singly-recurrent and non-recurrent outputs are used. No multiple recurrences.

### PR DESCRIPTION
**Pull request by Thrandis**
Fixes #5106

I made the changes to remove the limitation that the length of the sequence has to be a multiple of `save_every_N`. It still requires proper testing, and also I still need to check the behavior when the user pases `n_steps`.

### CODE REVIEW

**Comment by Thrandis** [comment]
@lamblin @nouiz Is there a way to check that `n_steps > 0` when `n_steps` is symbolic? Because it may break the padding otherwise...

**Comment by nouiz** [comment]
We could add an assert in the graph, but this have some inconvenience.

What happen when n_steps is negative? If it raise an error, even not very clear, it is fine for me. We don't support that for a long time. Just add a comment in the code where the reshape is added.

File: theano/scan_module/scan_checkpoints.py [hunk 1]
```diff
@@ -108,8 +105,10 @@ def scan_checkpoints(fn, sequences=[], outputs_info=None, non_sequences=[],
     o_n_steps = theano.tensor.cast(theano.tensor.ceil(n_steps / save_every_N),
                                    'int64')
     i_n_steps = save_every_N * theano.tensor.ones((o_n_steps,), 'int64')
-    i_n_steps = theano.tensor.inc_subtensor(i_n_steps[-1],
-                                            - n_steps % save_every_N)
+    mod = n_steps % save_every_N
```
Comment (nouiz) [inline theano/scan_module/scan_checkpoints.py]: If you add support for neg value, test it.

**Comment by Thrandis** [comment]
@nouiz I changed the `padding` flag.

For the `n_steps`, it will cause the `scan` to not behave properly, so it would be nice to make sure it is positive. But if it is the case also for the "classic" `scan`, we could do the check there.

**Comment by nouiz** [comment]
I think the normal scan raise an error for negative n step. Can you double check that?

**Comment by Thrandis** [comment]
@nouiz no, if `n_steps` is negative, `scan` will go backwards.

**Comment by nouiz** [comment]
The execution engine don't support it anymore:

https://github.com/Theano/Theano/blob/master/theano/scan_module/scan_op.py#L1142

How did you tested that? With scan_with_checkpoints or just scan?

**Comment by Thrandis** [comment]
@nouiz I only looked at the theano function `scan`, not its internals actually.

So everything is fine then. Can you relaunch travis, it seems to have failed somehow.

File: theano/scan_module/scan_checkpoints.py [hunk 2]
```diff
@@ -96,16 +101,29 @@ def scan_checkpoints(fn, sequences=[], outputs_info=None, non_sequences=[],
...
 
-    # Compute the number of steps of the inner and of the outer scan
-    o_n_steps = theano.tensor.cast(n_steps / save_every_N, 'int64')
-    i_n_steps = save_every_N
+    # Compute the number of steps of the outer scan
+    o_n_steps = theano.tensor.cast(theano.tensor.ceil(n_steps / save_every_N),
+                                   'int64')
+    i_n_steps = save_every_N * theano.tensor.ones((o_n_steps,), 'int64')
+    mod = n_steps % save_every_N
+    sign = theano.tensor.sgn(mod)
```
Comment (nouiz) [inline theano/scan_module/scan_checkpoints.py]: We do not support negative steps in scan anymore. So go back to the simplest earlisest version. Where did you saw in the doc that we support negative steps? We should remove that from the doc.

Comment (Thrandis) [inline theano/scan_module/scan_checkpoints.py]: @nouiz So this is not to support `n_steps` negative. This is to compute how many steps the last inner scan must run.

And for the negative steps, it is in the docstring of `scan.py`

**Pull request merged**